nano/secure/ledger.cpp

```diff
@@ -1039,57 +1040,90 @@ void nano::ledger::dump_account_chain (nano::account const & account_a, std::ost
+{
+	auto dependencies (dependent_blocks (transaction_a, block_a));
+	return std::all_of (dependencies.begin (), dependencies.end (), [this, &transaction_a](nano::block_hash const & hash_a) {
+		return hash_a.is_zero () || store.block_exists (transaction_a, hash_a);
+	});
+}
+
+bool nano::ledger::can_vote (nano::transaction const & transaction_a, nano::block const & block_a)
+{
+	auto dependencies (dependent_blocks (transaction_a, block_a));
+	return std::all_of (dependencies.begin (), dependencies.end (), [this, &transaction_a](nano::block_hash const & hash_a) {
+		auto result (hash_a.is_zero ());
+		if (!result)
+		{
+			result = false;
+			auto block (store.block_get (transaction_a, hash_a));
+			if (block != nullptr)
+			{
+				nano::confirmation_height_info height;
+				auto error = store.confirmation_height_get (transaction_a, block->account ().is_zero () ? block->sideband ().account : block->account (), height);
+				debug_assert (!error);
+				result = block->sideband ().height <= height.height;
+			}
+		}
+		return result;
+	});
+}
+
+bool nano::ledger::is_epoch_link (nano::link const & link_a)
+{
+	return network_params.ledger.epochs.is_epoch_link (link_a);
+}
+
+class dependent_block_visitor : public nano::block_visitor
 {
 public:
-	block_fit_visitor (nano::ledger & ledger_a, nano::transaction const & transaction_a) :
+	dependent_block_visitor (nano::ledger & ledger_a, nano::transaction const & transaction_a) :
 	ledger (ledger_a),
 	transaction (transaction_a),
-	result (false)
+	result ({ 0, 0 })
 	{
 	}
 	void send_block (nano::send_block const & block_a) override
 	{
-		result = ledger.store.block_exists (transaction, block_a.previous ());
+		result[0] = block_a.previous ();
 	}
 	void receive_block (nano::receive_block const & block_a) override
 	{
-		result = ledger.store.block_exists (transaction, block_a.previous ());
-		result &= ledger.store.block_exists (transaction, block_a.source ());
+		result[0] = block_a.previous ();
+		result[1] = block_a.source ();
 	}
 	void open_block (nano::open_block const & block_a) override
 	{
-		result = ledger.store.block_exists (transaction, block_a.source ());
+		if (block_a.source () != ledger.network_params.ledger.genesis_account)
+		{
+			result[0] = block_a.source ();
+		}
 	}
 	void change_block (nano::change_block const & block_a) override
 	{
-		result = ledger.store.block_exists (transaction, block_a.previous ());
+		result[0] = block_a.previous ();
 	}
 	void state_block (nano::state_block const & block_a) override
 	{
-		result = block_a.previous ().is_zero () || ledger.store.block_exists (transaction, block_a.previous ());
-		if (result && !ledger.is_send (transaction, block_a))
+		result[0] = block_a.hashables.previous;
+		result[1] = block_a.hashables.link;
+		if (ledger.is_epoch_link (block_a.hashables.link) || (ledger.store.block_exists (transaction, block_a.hashables.previous) && ledger.is_send (transaction, block_a)))
```

Comment:
I think we need some fail safe version of nano::ledger::is_send () function to save 1 read here while most blocks should have sideband attached

Reply:
Nice!

Added a debug_assert to `is_send` in https://github.com/nanocurrency/nano-node/pull/2785/commits/69f4818e84b193443355cd8b09f34709210a1610 similar to `block_source`

The easiest way to save the I/O read imo is an early check that the block has sideband, this is done in https://github.com/nanocurrency/nano-node/pull/2785/commits/a5f8eed2129d75526c2e4c971d96c86c899de947